src/mlpack/methods/ann/ffn_impl.hpp

```diff
@@ -494,23 +494,23 @@ typename MatType::elem_type FFN<
   // pass.
   networkOutput.set_size(network.OutputSize(), batchSize);
 
-  network.Forward(predictors.cols(begin, begin + batchSize - 1), networkOutput);
+  network.Forward(MatType(predictors.cols(begin, begin + batchSize - 1), true), networkOutput);
```

Comment:
Use of this MatType constructor is strongly not advised, as it's part of internal Armadillo functionality (ie. the subview and bool arguments).

Mat constructors that are part of the public API are listed in the documentation: https://arma.sourceforge.net/docs.html#Mat

Don't use any other constructors.  Armadillo API policy is explicitly documented here: https://arma.sourceforge.net/docs.html#changelog

Extract: "(...) Any functionality within Armadillo which is not explicitly described in the public API documentation is considered as internal implementation details, and may be changed or removed without notice."

Reply:
Sure, understood.  I guess the quasi_unwrap class is out of bounds too then?  I could use aliasing, as is done in other parts of the code, by doing the pointer arithmetic to pick out the appropriate columns, but this seems less safe than using classes/methods which have been written explicitly to do this.  

The goal is to prevent the unnecessary copying of potentially very large amounts of data.  Obviously armadillo can't make any assumptions about the lifetime of the referenced matrix, which is why they make this off-limits, but would be nice if we could explicitly call it if we have knowledge of this (the way they allow the memptr version).

Reply:
This is a nice find that I overlooked during the refactoring of the `FFN` code some time back.  I believe my original intention was that each layer would operate on the subviews directly, since the layers are templated---but this clearly does not work since the `Forward()` and `Backward()` and `Gradient()` functions are not templatized on their arguments, but instead at class level (and the type will be, e.g., `arma::mat`, not some subview or other arbitrary expression).

So, I think `MakeAlias()` with `.colptr()` is the right way to go here.

Thanks for finding this bug!  Definitely we don't want to make copies, that is a huge priority.